src/main/java/co/cask/spanner/source/SpannerSource.java

```diff
@@ -0,0 +1,266 @@
+          builder.set(fieldName, resultSet.getDouble(fieldName));
+          break;
+        case STRING:
+          builder.set(fieldName, resultSet.getString(fieldName));
+          break;
+        case BYTES:
+          ByteArray byteArray = resultSet.getBytes(fieldName);
+          builder.set(fieldName, byteArray.toByteArray());
+          break;
+      }
+    }
+    emitter.emit(builder.build());
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    LOG.info("Run finished, closing spanner client");
+    // free up spanner resources
+    if (spanner != null) {
+      spanner.close();
+    }
+  }
+
+  private void initializeConfig(Configuration configuration) {
+    setIfValueNotNull(configuration, SpannerConstants.PROJECT_ID, config.project);
+    setIfValueNotNull(configuration, SpannerConstants.SERVICE_ACCOUNT_FILE_PATH, config.serviceAccountFilePath);
+    setIfValueNotNull(configuration, SpannerConstants.INSTANCE_ID, config.instance);
+    setIfValueNotNull(configuration, SpannerConstants.DATABASE, config.database);
+    setIfValueNotNull(configuration, SpannerConstants.QUERY, String.format("Select * from %s;", config.table));
+  }
+
+  private void setIfValueNotNull(Configuration configuration, String key, String value) {
+    if (value != null) {
+      configuration.set(key, value);
+    }
+  }
+
+  /**
+   * Serialize the object into bytes and encode the bytes into string using Base64 encoder.
+   * @throws IOException
+   */
+  private String getSerializedObjectString(Object object) throws IOException {
+    try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+         ObjectOutputStream objectOutputStream = new ObjectOutputStream(byteArrayOutputStream)) {
+      objectOutputStream.writeObject(object);
+      objectOutputStream.flush();
+      return Base64.getEncoder().encodeToString(byteArrayOutputStream.toByteArray());
+    }
+  }
+
+  private PartitionOptions getPartitionOptions() {
+    PartitionOptions.Builder builder = PartitionOptions.newBuilder();
+    if (config.partitionSizeMB != null) {
+      builder.setPartitionSizeBytes(config.partitionSizeMB * 1024 * 1024);
+    }
+    if (config.maxPartitions != null) {
+      builder.setMaxPartitions(config.maxPartitions);
+    }
+    return builder.build();
+  }
+
+  /**
+   * Endpoint method to get the output schema of spanner query.
+   *
+   * @param request {@link SpannerSourceConfig} containing information required for query to execute.
+   * @return schema of fields
+   */
+  @Path("getSchema")
+  public Schema getSchema(SpannerSourceConfig request) throws Exception {
+    Spanner spanner = SpannerUtil.getSpannerService(request.serviceAccountFilePath, request.project);
+    DatabaseClient databaseClient =
+      spanner.getDatabaseClient(DatabaseId.of(request.project, request.instance, request.database));
+    String query = String.format("Select * from %s LIMIT 1;", request.table);
```

Comment:
Does spanner have any classes that can be used to build a query? Creating queries like this is generally not a good idea due to sql injection.

Reply:
`Statement` class is used to build SQL statement to execute. http://googlecloudplatform.github.io/google-cloud-java/google-cloud-clients/apidocs/com/google/cloud/spanner/Statement.html, the only improvement we can do is to use parameter binding to bind the table name, but even that doesn't provide any additional protection.

Reply:
these types of classes generally perform checks to prevent injection